projects/components/src/filter-button/filter-button.service.ts

```diff
@@ -0,0 +1,56 @@
+import { Injectable } from '@angular/core';
+import { NavigationService } from '@hypertrace/common';
+import { FilterAttribute } from '../filter-bar/filter-attribute';
+import { FilterBuilderService } from '../filter-bar/filter/builder/filter-builder.service';
+import { Filter } from '../filter-bar/filter/filter-api';
+import { FilterParserService } from '../filter-bar/filter/parser/filter-parser.service';
+
+@Injectable({ providedIn: 'root' })
+export class FilterButtonService {
+  private static readonly FILTER_QUERY_PARAM: string = 'filter';
+
+  public constructor(
+    private readonly navigationService: NavigationService,
+    private readonly filterBuilderService: FilterBuilderService,
+    private readonly filterParserService: FilterParserService
+  ) {}
+
+  public buildAvailableFilters(attribute: FilterAttribute, value: unknown): Filter[] {
+    return this.filterBuilderService.lookup(attribute).buildFiltersForAvailableOperators(attribute, value);
+  }
+
+  /*
+   * TODO: Combine service with filter-bar.service. A few duplicated methods here.
+   */
+
+  public applyUrlFilter(attributes: FilterAttribute[], filter: Filter): void {
```

Comment:
If we can remove attributes from this API (which makes sense I would think, we don't need attributes to convert to a url param) then I think we can clean up a bunch - including tying attributes into the table directly. So as I read this we need these to read in the current filters, replace if a field match, and otherwise append. But we could probably write some logic to do the same without doing a full parse (e.g. split on any operator and convert to a KV map, where K is the source name, V is the original url string)

Reply:
I can see the argument from a cleanliness perspective, but I'm not keen on making the presence of an object run dual purpose and also mean we want to filter. It could make the API confusing.

I'll make this change for now, but I'm posting this so I can link to it some day in the future. :-)

Reply:
This sounds like a response to https://github.com/hypertrace/hypertrace-core-ui/pull/75#discussion_r468958050 ?

If so, we could tweak the name from attribute to `filterAttribute` or something, but what I want to avoid is having separate inputs saying should I do something and how to do it - unless it's something that is meaningful to support changing at runtime and the like, such as column visibility. If you tell me (me being the table) how to filter, then I should expose it and if you don't tell me how, I shouldn't. The separation leads to either an unnecessary error conditions where you mark something as filterable but don't have an attribute, or extra data when you tell me how to filter but don't use it.

That said, my original comment above about removing the attributes coming in to table I think is one of the more important ones on the review so let's revisit that one.

Reply:
> As for removing attributes from table, I think that's just too much right now. All of our filtering code depends on the metadata and I think our time is better spent elsewhere than rewriting that for this PR. 

I'd agree that we shouldn't rewrite that here, but I don't think that's what would be needed to remove this. Unless I'm misunderstanding which is entirely possible, the only place this is used is in replacing an existing filter in the URL. We could almost certainly do a trivial parse (without touching the existing logic) to accomplish that which would not require attributes. More importantly - I'm not convinced that's the right behavior. If I have a filter like duration > 10, and add another filter like duration <15, those should both be applicable. Why are we trying to enforce a single filter per key?

Reply:
Also (just to pin this convo into GH so we can revisit it later if needed) if we were to use a provider for communication rather than the URL (besides making this capability more portable to work in places we're not using a url sync'd table), the need to parse would go away and we'd also be able to avoid the attribute input.

Reply:
I guess I would think that being more lax with the filters would be a better starting point as it 
1. Allows more functionality (not just things like ranges, but also `!= x and != y`, or `like x and != xy`)
2. Is simpler to implement - we have code existing to enforce this right now, including this thread
3. Does allow users to create nonsensical queries, but that's always possible (e.g. I could query for status = success and statusCode = 500). I think that's the trade off for flexibility. The important thing is to make the query obvious. If we want to add certain rules to prevent knowably wrong things as an improvement, I think that can come iteratively (for example not allowing a second equality filter for the same field).

Reply:
Okay. I'll take care of it with the other jira.